sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,290 @@
+  public void sendBufferedDataAndFinishOutboundStreams() {
+    LOG.debug("Closing streams for outbound endpoints {}", outputLocations);
+    LOG.debug("Sent outbound data size : {}", perEndpointByteCount);
+    LOG.debug("Sent outbound element count : {}", perEndpointElementCount);
+    if (byteCounter.get() == 0 && outputLocations.isEmpty()) {
+      return;
+    }
+    Elements.Builder bufferedElements = convertBufferForTransmission();
+    for (LogicalEndpoint outputLocation : outputLocations.keySet()) {
+      if (outputLocation.isTimer()) {
+        bufferedElements
+            .addTimersBuilder()
+            .setInstructionId(outputLocation.getInstructionId())
+            .setTransformId(outputLocation.getTransformId())
+            .setTimerFamilyId(outputLocation.getTimerFamilyId())
+            .setIsLast(true);
+      } else {
+        bufferedElements
+            .addDataBuilder()
+            .setInstructionId(outputLocation.getInstructionId())
+            .setTransformId(outputLocation.getTransformId())
+            .setIsLast(true);
+      }
+    }
+    outboundObserver.onNext(bufferedElements.build());
+    outputLocations.clear();
+    perEndpointByteCount.clear();
+    perEndpointElementCount.clear();
+  }
+
+  public <T> void acceptData(String pTransformId, T data) throws Exception {
+    accept(LogicalEndpoint.data(processBundleRequestIdSupplier.get(), pTransformId), data);
+  }
+
+  public <T> void acceptTimers(String pTransformId, String timerFamilyId, T timers)
+      throws Exception {
+    accept(
+        LogicalEndpoint.timer(processBundleRequestIdSupplier.get(), pTransformId, timerFamilyId),
+        timers);
+  }
+
+  public <T> void accept(LogicalEndpoint endpoint, T data) throws Exception {
+    if (timeLimit > 0) {
+      checkFlushThreadException();
+    }
+    buffer.compute(
+        endpoint,
+        (e, output) -> {
+          if (output == null) {
+            output = ByteString.newOutput();
+          }
+          int size = output.size();
+          try {
+            ((Coder<T>) outputLocations.get(e)).encode(data, output);
+          } catch (IOException ex) {
+            throw new RuntimeException("Failed to encode data.");
+          }
+          if (output.size() - size == 0) {
+            output.write(0);
+          }
+          final long delta = (long) output.size() - size;
+          byteCounter.getAndAdd(delta);
+          perEndpointByteCount.compute(
+              e, (e1, byteCount) -> byteCount == null ? delta : delta + byteCount);
+          return output;
+        });
+
+    if (byteCounter.get() >= sizeLimit) {
+      flush();
+    }
+  }
+
+  private Elements.Builder convertBufferForTransmission() {
+    Elements.Builder bufferedElements = Elements.newBuilder();
+    for (LogicalEndpoint endpoint : buffer.keySet()) {
+      ByteString.Output output = buffer.remove(endpoint);
+      if (output == null) {
+        continue;
+      }
+      byteCounter.getAndAdd(-1 * output.size());
```

Comment:
sum up all your subtractions locally and only do one atomic update at the end of the method.